src/sunmatrix/sparse/sunmatrix_sparse.c

```diff
@@ -724,30 +745,39 @@ int SUNMatScaleAddI_Sparse(realtype c, SUNMatrix A)
       /* set current column (row) pointer to current # nonzeros */
       Cp[j] = nz;
 
-      /* clear out temporary arrays for this column (row) */
-      for (i=0; i<M; i++) {
-        w[i] = 0;
-        x[i] = 0.0;
-      }
+      /* clear out diagonal element for this column (row) */
+      x[j] = RCONST(0.0);
 
       /* iterate down column (along row) of A, collecting nonzeros */
       for (p=Ap[j]; p<Ap[j+1]; p++) {
-        w[Ai[p]] += 1;         /* indicate that row is filled */
         x[Ai[p]] = c*Ax[p];    /* collect/scale value */
       }
 
       /* add identity to this column (row) */
       if (j < M) {
-        w[j] += 1;     /* indicate that row is filled */
         x[j] += ONE;   /* update value */
       }
 
       /* fill entries of C with this column's (row's) data */
-      for (i=0; i<M; i++) {
-        if ( w[i] > 0 ) {
-          Ci[nz] = i;
-          Cx[nz++] = x[i];
-        }
+      for (p=Ap[j]; p<Ap[j+1] && Ai[p]<j; p++) {
+        Ci[nz] = Ai[p];
+        Cx[nz++] = x[Ai[p]];
+      }
+      if (Ai[p] != j || Ap[j] == Ap[j+1]) {
+        Ci[nz] = j;
+        Cx[nz++] = x[j];
+      }
+      for (; p<Ap[j+1]; p++) {
```

Comment:
Again, I recommend small comments be added to clarify structure, e.g.:
```suggestion
      /*  fill entries before diagonal */
      for (p=Ap[j]; p<Ap[j+1] && Ai[p]<j; p++) {
        Ci[nz] = Ai[p];
        Cx[nz++] = x[Ai[p]];
      }
      /*  fill diagonal if applicable */
      if (Ai[p] != j || Ap[j] == Ap[j+1]) {
        Ci[nz] = j;
        Cx[nz++] = x[j];
      }
      /*  fill entries after diagonal */
      for (; p<Ap[j+1]; p++) {
```